File: Q1_Cpp/src/Product/Product.cpp
#include "../../include/Product/Product.h"

// Constructor to initialize a Product
Product::Product(std::string id, std::string name, Supplier supplier, int packagingUnits, double price) 
    : id(id), name(name), supplier(supplier), packagingUnits(packagingUnits), price(price) {
}

// Getter and setter methods implementation
std::string Product::getId() const { return id; }
void Product::setId(const std::string& id) { this->id = id; }
std::string Product::getName() const { return name; }
void Product::setName(const std::string& name) { this->name = name; }
Supplier Product::getSupplier() const { return supplier; }
void Product::setSupplier(const Supplier& supplier) { this->supplier = supplier; }
int Product::getPackagingUnits() const { return packagingUnits; }
void Product::setPackagingUnits(int packagingUnits) { this->packagingUnits = packagingUnits; }
double Product::getPrice() const { return price; }
void Product::setPrice(double price) { this->price = price; }

// Note: The abstract methods 'calculateDiscount' and 'displayProductInfo' are not implemented here
// They should be implemented in the derived classes

File: Q1_Java/src/main/java/Product/Product.java
package Product; // Package for the class

public abstract class Product {
    private String id; // Product's ID
    private String name; // Product's name
    private Supplier supplier; // Supplier object for the product
    private int packagingUnits; // Packaging units of the product
    private double price; // Product's price

    // Constructor to initialize a Product
    public Product(String id, String name, Supplier supplier, int packagingUnits, double price) {
        this.id = id; // Setting product ID
        this.name = name; // Setting product name
        this.supplier = supplier; // Setting product's supplier
        this.packagingUnits = packagingUnits; // Setting packaging units
        this.price = price; // Setting price
    }

    // Abstract methods to be implemented in subclasses
    public abstract double calculateDiscount(int quantity, int month);
    public abstract String displayProductInfo();

    // Getter and setter methods
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public Supplier getSupplier() { return supplier; }
    public void setSupplier(Supplier supplier) { this.supplier = supplier; }
    public int getPackagingUnits() { return packagingUnits; }
    public void setPackagingUnits(int packagingUnits) { this.packagingUnits = packagingUnits; }
    public double getPrice() { return price; }
    public void setPrice(double price) { this.price = price; }
}
